examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,465 @@
+    PetscCall(VecGetLocalSize(U_loc, &num_mesh_points_loc));
+    PetscCall(VecGetArray(U_loc, &u_array));
+    PetscCall(VecGetArrayRead(X_loc, &x_array));
+    for (PetscInt i = 0; i < num_mesh_points_loc; i++) {
+      PetscScalar x[dim];
+
+      for (PetscInt d = 0; d < dim; d++) x[d] = x_array[d * num_mesh_points_loc + i];
+      u_array[i] = EvalU(dim, x);
+    }
+    PetscCall(VecRestoreArray(U_loc, &u_array));
+    PetscCall(VecRestoreArrayRead(X_loc, &x_array));
+    PetscCall(VecZeroEntries(U_mesh));
+    PetscCall(DMLocalToGlobal(dm_mesh, U_loc, INSERT_VALUES, U_mesh));
+    PetscCall(DMRestoreLocalVector(dm_mesh, &U_loc));
+  }
+
+  // Interpolate from mesh to points via PETSc
+  {
+    PetscDS            ds;
+    PetscFE            fe;
+    PetscFEGeom        fe_geometry;
+    PetscQuadrature    quadrature;
+    Vec                U_loc;
+    PetscInt           cell_start, cell_end;
+    PetscScalar       *u_all_points_array;
+    const PetscScalar *coord_points;
+
+    PetscCall(DMGetLocalVector(dm_mesh, &U_loc));
+    PetscCall(VecZeroEntries(U_loc));
+    PetscCall(DMGlobalToLocal(dm_mesh, U_mesh, INSERT_VALUES, U_loc));
+
+    PetscCall(DMGetDS(dm_mesh, &ds));
+    PetscCall(PetscDSGetDiscretization(ds, 0, (PetscObject *)&fe));
+    PetscCall(DMSwarmSortGetAccess(dm_swarm));
+    PetscCall(DMPlexGetHeightStratum(dm_mesh, 0, &cell_start, &cell_end));
+    PetscCall(DMSwarmGetField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&coord_points));
+    PetscCall(DMSwarmGetField(dm_swarm, "u", NULL, NULL, (void **)&u_all_points_array));
+
+    // -- Interpolate values to each swarm point, one element in the background mesh at a time
+    for (PetscInt cell = cell_start; cell < cell_end; cell++) {
+      PetscTabulation tabulation;
+      PetscScalar    *u_cell = NULL, *coords_points_true, *coords_points_ref;
+      PetscReal       v[dim], J[dim * dim], invJ[dim * dim], detJ;
+      PetscInt       *points;
+      PetscInt        num_points_in_cell;
+
+      PetscCall(DMSwarmSortGetPointsPerCell(dm_swarm, cell, &num_points_in_cell, &points));
+      PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_true));
+      PetscCall(DMGetWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_ref));
+      // ---- Reference coordinates for swarm points in background mesh element
+      for (PetscInt p = 0; p < num_points_in_cell; p++) {
+        for (PetscInt d = 0; d < dim; d++) coords_points_true[p * dim + d] = coord_points[points[p] * dim + d];
+      }
+      PetscCall(DMPlexCoordinatesToReference(dm_mesh, cell, num_points_in_cell, coords_points_true, coords_points_ref));
+
+      // ---- Interpolate values from current element in background mesh to swarm points
+      PetscCall(PetscFECreateTabulation(fe, 1, num_points_in_cell, coords_points_ref, 1, &tabulation));
+      PetscCall(DMPlexComputeCellGeometryFEM(dm_mesh, cell, NULL, v, J, invJ, &detJ));
```

Comment:
we aren't using the outputs of this call, that seems fishy

Reply:
That's a placeholder for the grad checking code